### ISSUE TITLE
The more classes you have instantiated, the longer it takes to instantiate another one.

### ISSUE THREAD

**Issue by niklwors** [OpenModelica/OpenModelica#10722]
### Description

With the new instantiation-based GUI it takes longer and longer to instantiate another class. This is especially the case for larger models.

### Steps to Reproduce

- Create a new model and insert the component ```Rexroth_BRSL.Generic.Hydraulics.Sources.Tank``` from the BRSL library.
- This takes about one second.
- Add the model ```Rexroth_BRSL.Examples.Systems.SingleAxis``` to sheet.
- Add once more the component ```Rexroth_BRSL.Generic.Hydraulics.Sources.Tank``` to the model.
- Now it already takes 14s.

### Version and OS

 - OpenModelica Version: v1.22.0-dev-105-g997b5e0116
 - OS: Windows 10, 64 bit
 - BRSL 2.3..0

**Comment by casella** [comment]
@adeas31 this looks potentially serious, please have a look at it ASAP.

**Comment by casella** [comment]
@adeas31, @perost, I am still wondering why this `SingleAxis` model should contain so much information (40 MB worth of JSON stuff, according to this [comment](https://github.com/OpenModelica/OpenModelica/issues/10436#issuecomment-1482662610)) that it takes 15 seconds to parse it with a 4 GHz CPU. At the end of the day, we only need to show a 100x100 pixel icon and a handful of parameters with their comments. I am positive that more than 99% of the information contained in that JSON file is totally irrelevant.

Improving the peformance of JSON reading and writing is fine with me, but I'm not sure making it faster to shuttle large amount of irrelevant information back and forth is the best use of our scarce resources 😅.

Can you carry out a bit of analysis to figure out what parts of that JSON are certainly irrelevant and could be left out? Maybe we are still passing around protected stuff? Model equations? Or other types of information that are irrelevant for graphical rendering and parameter editing?

If you can post that file here (maybe zipped), I can also have a look at it myself.

Thanks!

**Comment by casella** [comment]
> We dump everything that OMEdit needs, not just for drawing the diagrams but also for e.g. parameter editing, replaceable, and so on

Sure, but it can't be 40 MB.

> But a lot of the time is spent on actually instantiating the model, which was always going to be an issue for large models, and then creating the data structures in OMEdit from the JSON.

Do we have a breakdown of how much each of these steps takes?

**Comment by adeas31** [comment]
> Do we have a breakdown of how much each of these steps takes?

Yes. See https://github.com/OpenModelica/OpenModelica/issues/10436#issuecomment-1482662610

**Comment by adeas31** [comment]
9.738 is the time for instantiation. The total time is the sum of all i.e., 0.37 + 6.649 + 0.111 + 9.738 = 16.8

Also note that this varies based on the OS. On Windows it takes more time.

**Comment by perost** [comment]
#10743 improves the performance of the "parsing JSON" phase in OMEdit (of which actual parsing is a negligible part), reducing the time by about half (so from 6s to 3s for SingleAxis). There's probably more that can be done, I haven't run the profiler again yet, but this takes care of some of the low-hanging fruit at least.

**Comment by adeas31** [comment]
#10753 further improves JSON parsing.

**Comment by casella** [comment]
Maybe we could do some profiling again after the latest commits and see what the current bottlenecks are. 

@adeas31 could you take care of that?

**Comment by perost** [comment]
#10919 further improves the instantiation of the BRSL models, and any libraries with a lot of heavy outer elements. For the SingleAxis model it improves the performance of `getModelInstance` from about 8s to 2.7s on my computer.

Before (note that the time doesn't start from 0 since I'm not including the time it takes to load the libraries):
```
Notification: Performance of NFApi.mkTop(Rexroth_BRSL.Examples.Systems.SingleAxis): time 0.236/1.89, allocations: 205 MB / 0.8775 GB, free: 9.801 MB / 0.7607 GB
Notification: Performance of Inst.instantiateRootClass: time 4.819/6.708, allocations: 2.858 GB / 3.736 GB, free: 12.57 MB / 2.073 GB
Notification: Performance of NFApi.buildInstanceTree: time 0.0339/6.742, allocations: 7.411 MB / 3.743 GB, free: 7.793 MB / 2.073 GB
Notification: Performance of Inst.instExpressions: time 1.102/7.844, allocations: 380.1 MB / 4.114 GB, free: 12.08 MB / 2.354 GB
Notification: Performance of Typing.typeComponents: time 0.02931/7.874, allocations: 11.26 MB / 4.125 GB, free: 0.7188 MB / 2.354 GB
Notification: Performance of Typing.typeBinding: time 0.07532/7.949, allocations: 25.54 MB / 4.15 GB, free: 6.98 MB / 2.386 GB
Notification: Performance of NFApi.dumpJSONInstanceTree: time 1.695/9.644, allocations: 0.6814 GB / 4.832 GB, free: 9.773 MB / 2.776 GB
Notification: Performance of JSON.toString: time 0.4149/10.06, allocations: 112.5 MB / 4.942 GB, free: 39.67 MB / 2.918 GB
```
After:
```
Notification: Performance of NFApi.mkTop(Rexroth_BRSL.Examples.Systems.SingleAxis): time 0.2347/1.845, allocations: 205 MB / 0.8775 GB, free: 9.719 MB / 0.7607 GB
Notification: Performance of Inst.instantiateRootClass: time 0.5992/2.444, allocations: 367.9 MB / 1.237 GB, free: 11.88 MB / 1.026 GB
Notification: Performance of NFApi.buildInstanceTree: time 0.03289/2.477, allocations: 7.411 MB / 1.244 GB, free: 4.445 MB / 1.026 GB
Notification: Performance of Inst.instExpressions: time 0.1672/2.644, allocations: 57.83 MB / 1.301 GB, free: 10.93 MB / 1.089 GB
Notification: Performance of Typing.typeComponents: time 0.02908/2.673, allocations: 11.25 MB / 1.312 GB, free: 15.58 MB / 1.104 GB
Notification: Performance of Typing.typeBinding: time 0.07536/2.749, allocations: 25.54 MB / 1.336 GB, free: 5.844 MB / 1.12 GB
Notification: Performance of NFApi.dumpJSONInstanceTree: time 1.62/4.369, allocations: 0.6813 GB / 2.018 GB, free: 119 MB / 1.511 GB
Notification: Performance of JSON.toString: time 0.4064/4.776, allocations: 112.5 MB / 2.128 GB, free: 97.87 MB / 1.575 GB
```
It's mostly just the instantiation that's affected, since we didn't type outer components previously either. The bottleneck now is actually dumping the model to JSON, which seems a bit odd. I'll need to do some profiling and see why that is. One option could be to just skip the JSON entirely and pass the MetaModelica structure directly to OMEdit, but we'll see if that's needed or not.

**Comment by abuntrock** [comment]
As @niklwors mentioned in his email this morning to most of you OSMC developers, we at Bosch Rexroth released our BRSL 3.0 in the past days and are lucky to distribute it as compatible to OMEdit thanks to the many improvements over the past year. 

This is also means that we are slowly starting to get more feedback from Rexroth internal users, which are the ones actually building up the large (hydraulic) system models. The performance degradation when working with large models is still at a level that makes working with them quite unpleasant.

The problem is that "simple tasks" like **removing a connection line** can take very long in large models. 
Consider the example  `Rexroth_BRSL.Rexroth_MobileControls.Examples.Excavator_with_RCS35_Valve` which is part of the BRSL that @niklwors uploaded to his OMDev folder. Removing (or drawing) any connection in this model takes up to **1 minute** on my laptop which is quite new (presumably to all the json stuff getting send back and forth).

We just wanted to say that despite all the improvements regarding the new API over the last year we and our BRSL users are still looking forward towards any performance improvements especially when the actual task (deleting one connection) does not seem so complicated to the user. #10436 @perost

**Comment by casella** [comment]
See the plan in #11920. Removing components, adding and removing connections will be fast in 1.23.0. Adding new components will take more time, we now have a plan are working on that but it will be in 1.24.0, because it will need thorough testing.
